bin/runtime-common/src/messages.rs

```diff
@@ -1183,4 +1194,56 @@ mod tests {
 			.collect()),
 		);
 	}
+
+	#[test]
+	fn verify_messages_proof_with_parser_does_not_panic_if_messages_count_mismatches() {
+		assert_eq!(
+			target::verify_messages_proof_with_parser::<OnThisChainBridge, _, _>(
+				(
+					Default::default(),
+					StorageProof::new(vec![]),
+					Default::default(),
+					0,
+					u64::MAX
+				),
+				0,
+				|_, _| Ok(TestMessageProofParser {
+					failing: false,
+					messages: 0..=u64::MAX,
+					outbound_lane_data: Some(OutboundLaneData {
+						oldest_unpruned_nonce: 1,
+						latest_received_nonce: 1,
+						latest_generated_nonce: 1,
+					}),
+				}),
+			),
+			Err(target::MessageProofError::MessagesCountMismatch),
+		);
+	}
+
+	#[test]
+	#[should_panic]
```

Comment:
I'm not a fan of `should_panic` tests, had to actually run the code without this to understand that the panic is coming from `Vec::with_capacity(u64::max_value())` call (note the behavior is in theory platform specific).

TBH I'd prefer to explicitly disallow allocations over `u32::max_value()` and just return a regular error here. I know we do verify the incoming values using `MaxMessagesAtInboundLane` in practice, but I'd still be more in favor of defensive programming style here.

I'd simply change `messages_count: MessageNonce` parameter to `u32` (I'm assuming `MessageNonce: From<u32>` is implemented anyway) to make it explicit (so it doesn't even require documentation that much).